### PR TITLE
[Blazor] Consume fingerprinted assets in MVC and Blazor

### LINKED ISSUE #56076: [Blazor] Runtime APIs to support fingerprinting
# Description

These set of APIs allow Blazor, MVC and Razor pages to map well-known URLs to content-specific (fingerprinted URLs).

### Startup/Program APIs

The list of assets is defined on a per-endpoint basis. MVC and RazorPages opt-in to the feature via a call to `WithResourceCollection` following the call to `MapRazorPages()` or `MapControllers()`.

This call adds a `ResourceAssetCollection` to the Metadata collection of the endpoints, which contains the mapping between human-readable URLs and content-specific URLs.

From there, other components can retrieve the mapping from endpoint metadata.

```diff
app.MapStaticAssets();

app.MapRazorPages()
+   .WithResourceCollection();
```

Since we can have calls to map different manifests, `WithResourceCollection` takes a parameter to provide the Id for the matching `MatchStaticAssets`.

```diff
app.MapStaticAssets("WebassemblyApp1.Client");
app.MapStaticAssets("WebassemblyApp2.Client");

app.MapRazorPages()
+   .WithResourceCollection("WebassemblyApp1.Client");
```

### Consuming the mapped assets

The assets are consumed indirectly from MVC and Razor Pages via existing `Script`, `Image`, `Link`, and `Url` tag helpers.

In Blazor, the assets are consumed via the `Assets` property in `ComponentBase` which exposes an indexer to resolve the fingerprinted url for a given asset.

```html
<link rel="stylesheet" href="@Assets["app.css"]" />
<link rel="stylesheet" href="@Assets["BlazorWeb-CSharp.styles.css"]" />
```

This in the future will be cleaner with a compiler feature similar to the Url tag helper in MVC that transforms "~/app.css" into the above code inside the href attribute.

In addition to this, there are two built-in components to generate an importmap for scripts. See [here](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/script/type/importmap) for details.

**TL;DR: Creates a mapping for calls to `import` and script `type="module"` that optionally includes integrity information**. A sample:

```html
<script type="importmap">
  {
    "imports": {
      "square": "./module/shapes/square.js"
    },
    "integrity": {
      "./module/shapes/square.js": "sha384-oqVuAfXRKap7fdgcCY5uykM6+R9GqQ8K/uxy9rx7HNQlGYl1kPzQho1wx4JwY8wC"
    }
  }
</script>
```

There is an ImportMap blazor component
```html
<ImportMap />
```

and in MVC we extended the script tag helper to support it
```html
<script type="importmap"></script>
```

### Layering

I included this section to explain the reasons for some of the APIs below.

```mermaid
flowchart LR
    subgraph AspNet[ASP.NET Framework reference]
        StaticAssets[Static Assets]
        Endpoints[Components Endpoints SSR]
        MVCRp[Mvc and Razor Pages]
    end
    Components[Components]
    ComponentsWasm[Components Webassembly]
    ComponentsWebView[Components WebView]
    ComponentsWasm-->Components
    ComponentsWebView-->Components
    Endpoints-->Components
    Endpoints-->StaticAssets
    MVCRp-->Endpoints
```

**Two important things:**
 * Components doesn't have a reference to ASP.NET Core because it runs in other environments, and is what Razor Class Libraries target.
 * MVC and Razor Pages consume the SSR support for components.

With this in mind, information for the assets is exposed differently from StaticAssets and Components

```mermaid
flowchart TD
    subgraph AspNet[ASP.NET Framework reference]
        subgraph StaticAssets[Static Assets]
            StaticAssetDescriptor
            StaticAssetSelector
            StaticAssetProperty
            StaticAssetResponseHeader
        end
        Endpoints[Components Endpoints SSR]
        MVCRp[Mvc and Razor Pages]
    end
    subgraph Components[Components]
       ResourceAssetCollection
       ResourceAsset
       ResourceAssetProperty
       ImportMapDefinition
    end
    ComponentsWasm[Components Webassembly]
    ComponentsWebView[Components WebView]
    ComponentsWasm-->Components
    ComponentsWebView-->Components
    Endpoints-->Components
    Endpoints-->StaticAssets
    MVCRp-->Endpoints
```

* ResourceAsset and its related types represent Blazor's view of a resource, which only includes things that are useful when including that resource in markup.
* ResourceAssetDescriptor and its related types are the "HTTP" view of the file/endpoint that we mapped.

The Component endpoints assembly is the one responsible for mapping the `StaticAssetDescriptor`s into the `ResourceAssetCollection` so that Blazor and MVC can consume them.

#### Microsoft.AspNetCore.Components.dll

```diff
namespace Microsoft.AspNetCore.Components;

public abstract class ComponentBase : IComponent, IHandleEvent, IHandleAfterRender
{
+    protected ResourceAssetCollection Assets { get; }
}

public readonly struct RenderHandle
{
+    public ResourceAssetCollection Assets { get; }
}

public abstract partial class Renderer : IDisposable, IAsyncDisposable
{
+    protected internal virtual ResourceAssetCollection Assets { get; } = ResourceAssetCollection.Empty;
}

+public class ResourceAssetCollection : IReadOnlyList<ResourceAsset>
+{
+   public static readonly ResourceAssetCollection Empty = new([]);
+   public ResourceAssetCollection(IReadOnlyList<ResourceAsset> resources);
+   public string this[string key];
+   public bool IsContentSpecificUrl(string path);
+}

+public class ResourceAsset(string url, IReadOnlyList<ResourceAssetProperty>? properties)
+{
+   public string Url { get; } = url;
+   public IReadOnlyList<ResourceAssetProperty>? Properties { get; } = properties;
+}

+public class ResourceAssetProperty(string name, string value)
+{
+   public string Name { get; } = name;
+   public string Value { get; } = value;
+}
```

#### Microsoft.AspNetCore.Components.Endpoints

```diff 
namespace Microsoft.AspNetCore.Components;

+public class ImportMap : IComponent
+{
+    [CascadingParameter] public HttpContext? HttpContext { get; set; } = null;
+    [Parameter] public ImportMapDefinition? ImportMapDefinition { get; set; }
+}

+public class ImportMapDefinition
+{
+   public ImportMapDefinition(
+        IReadOnlyDictionary<string, string>? imports,
+        IReadOnlyDictionary<string, IReadOnlyDictionary<string, string>>? scopes,
+        IReadOnlyDictionary<string, string>? integrity)

+   public static ImportMapDefinition FromResourceCollection(ResourceAssetCollection assets);
+   public static ImportMapDefinition Combine(params ImportMapDefinition[] sources);

+   public IReadOnlyDictionary<string, string>? Imports { get; }
+   public IReadOnlyDictionary<string, IReadOnlyDictionary<string, string>>? Scopes { get; }
+   public IReadOnlyDictionary<string, string>? Integrity { get; }
+   public override string ToString();
+}

public static class RazorComponentsEndpointConventionBuilderExtensions
{
+    public static RazorComponentsEndpointConventionBuilder WithResourceCollection(
+    this RazorComponentsEndpointConventionBuilder builder,
+    string? manifestPath = null);
}
```

#### Assembly Microsoft.AspNetCore.WebUtilities

```diff
namespace Microsoft.AspNetCore.WebUtilities;

public class WebEncoders
{
+    Base64UrlEncode(System.ReadOnlySpan<byte> input, System.Span<char> output)
}
```

#### Assembly Microsoft.AspNetCore.Mvc.RazorPages

```diff
namespace Microsoft.AspNetCore.Builder;

+public static class PageActionEndpointConventionBuilderResourceCollectionExtensions
+{
+    static WithResourceCollection(this Microsoft.AspNetCore.Builder.PageActionEndpointConventionBuilder! builder, string? manifestPath = null): Microsoft.AspNetCore.Builder.PageActionEndpointConventionBuilder!
+}
```

#### Microsoft.AspNetCore.Mvc.ViewFeatures

```diff
namespace Microsoft.AspNetCore.Builder;

+public static class ControllerActionEndpointConventionBuilderResourceCollectionExtensions
+{
+    static WithResourceCollection(this Microsoft.AspNetCore.Builder.ControllerActionEndpointConventionBuilder! builder, string? manifestPath = null): Microsoft.AspNetCore.Builder.ControllerActionEndpointConventionBuilder!
+}
```

#### Assembly Microsoft.AspNetCore.Mvc.TagHelpers.dll

```diff
namespace Microsoft.AspNetCore.Mvc.TagHelpers;

public class ScriptTagHelper
{
+    public string Type { get; set; }
+    public ImportMapDefinition ImportMap { get; set; }
}
```

#### Assembly Microsoft.AspNetCore.StaticAssets.dll

```diff
namespace Microsoft.AspNetCore.StaticAssets;

+public sealed class StaticAssetDescriptor
+{
+    public required string Route { get; set; }
+    public required string AssetPath { get; set; }
+    public IReadOnlyList<StaticAssetSelector> Selectors { get; set; }
+    public IReadOnlyList<StaticAssetProperty> Properties { get; set; }
+    public IReadOnlyList<StaticAssetResponseHeader> ResponseHeaders { get; set; }
+}

+public sealed class StaticAssetResponseHeader(string name, string value)
+{
+    public string Name { get; } = name;
+    public string Value { get; } = value;
+}

+public sealed class StaticAssetProperty(string name, string value)
+{
+    public string Name { get; } = name;
+    public string Value { get; } = value;
+}

+public sealed class StaticAssetSelector(string name, string value, string quality)
+{
+   public string Name { get; } = name;
+   public string Value { get; } = value;
+    public string Quality { get; } = quality;
+}
```

```diff
namespace Microsoft.AspNetCore.StaticAssets.Infrastructure;

public static class StaticAssetsEndpointDataSourceHelper
{
+    public static bool HasStaticAssetsDataSource(IEndpointRouteBuilder builder, string? staticAssetsManifestPath = null);

+    public static IReadOnlyList<StaticAssetDescriptor> ResolveStaticAssetDescriptors(
        IEndpointRouteBuilder endpointRouteBuilder,
        string? manifestPath)
}
```

### PR DESCRIPTION
**Pull request by javiercn**
Fixes #56076

### Overview

Fingerprinting endpoints are defined as part of the build process. The fingerprinted endpoints contain a property "fingerprint" with the fingerprint for the file content. 
These endpoints end up mapped as StaticAsset endpoints and include cache headers to ensure that the content is cached for a long time.

The StaticAssets routerware exposes an IReadOnlyList<StaticAssetDescriptor> to consumers. Each static asset descriptor represents an addressable resource and includes the route, the response headers, a set of properties that other layers can use to identify concrete assets, and a list of selectors, used to discriminate between resources mapped to the same path.

A `StaticAssetDescriptor` represents the "physical" representation of the file at the HTTP level.

The information from the descriptors is exposed to Components via the ResourceAssetCollection. This concept lives in Components and provides additional information about the resources in the app that can be used by components to obtain content specific URLs and to read additional properties from assets, like their integrity value, which can then be used to apply things like CSP or subresource integrity.

For scenarios see the companion issue #56076 and [this](https://github.com/dotnet/aspnetcore/pull/56045/commits/4a66520815de2ddfee7dcd4229c10fa9a2b33abb) commit that focuses on the template updates.

### Key design points

**Exposing the asset information to components**

This could have been done in three ways:
* Cascading Value.
* Scoped service.
* Through the RenderHandle

I chose to expose this through the render handle directly for a couple of motives:
* It's something deeply coupled to the application being rendered.
* We want this to be as performant as possible (ideally a dictionary lookup) and with minimal overhead.
* We want to work for all components.

Why I avoided a cascading value:
* Adding it as a cascading value requires the user to add a cascading value to their component every time, they want to look up an asset.
* This is not really a hierarchical concept, is a per-renderer concept, we don't want situations in which this can be replaced/overriden.
* Requires additional initialization on a per-request basis.

Why I avoided a scoped service:
* Requires initializing the service on every request.
* Does not work with OwningComponentBase components as the service will not be initialized.
* Requires injecting the service, which is additional overhead.

**How webassembly consumes fingerprinted assets**:
* When we detect webassembly or automodes are enabled for an app, we map an additional endpoint '_framework/resource-collection.<<fingerprint>>.js' which exposes the resource collection as a JS module.
* We emit the URL to the body of the request only when a webassembly component is rendered into the page as persisted component state.
* During WebAssembly boot, we retrieve the url, import the module and call a function to retrieve the asset collection and reconstruct it in memory.
* The URL we provide is content specific and is cached forever, so we only pay this cost once per user until the app gets updated.
* The resource collection is also exposed at a human-readable URL '_framework/resource-collection.js' so the JS side has access to it should it be needed (for enhanced nav or other frameworks, 3rd party components).

**Hot reload**
* During development/hot reload we disable certain features to ensure a good developer experience even in the event that some files change.
  * We remove the integrity information from the assets to avoid issues when a file is changed while the app is running.
  * We disable immutable caching from content-specific URLs and replace it with no-cache (in fact, we `no-cache` every asset/endpoint) to ensure that the browser always tries to retrieve up to date content.
    * As a note, we already support updating ETags and other properties on the flight.

**Integrations**
* For MVC and Razor pages, this is almost transparent, since it's baked in on their built-in tag helpers.
* For Blazor there is a new ImportMap component that takes care of rendering the importmap script.

### CODE REVIEW

File: src/Components/Components/src/ResourceAsset.cs [hunk 1]
```diff
@@ -0,0 +1,29 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Linq;
+
+namespace Microsoft.AspNetCore.Components;
+
+/// <summary>
+/// A resource of the component application, such as a script, stylesheet or image.
```
Comment (MackinnonBuck) [inline src/Components/Components/src/ResourceAsset.cs:10]:
```suggestion
/// A resource of the components application, such as a script, stylesheet or image.
```

File: src/Components/Components/src/ResourceAssetCollection.cs [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+        _uniqueUrlMappings = new Dictionary<string, ResourceAsset>(StringComparer.OrdinalIgnoreCase);
+        _contentSpecificUrls = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+        _resources = resources;
+        foreach (var resource in resources)
+        {
+            foreach (var property in resource.Properties ?? [])
+            {
+                if (property.Name.Equals("label", StringComparison.OrdinalIgnoreCase))
+                {
+                    _uniqueUrlMappings[property.Value] = resource;
```
Comment (MackinnonBuck) [inline src/Components/Components/src/ResourceAssetCollection.cs:37]: Would we consider throwing if two resource assets specify the same label?

File: src/Components/Components/src/ParameterView.cs [hunk 3]
```diff
@@ -46,6 +46,8 @@ private ParameterView(in ParameterViewLifetime lifetime, RenderTreeFrame[] frame
 
     internal ParameterViewLifetime Lifetime => _lifetime;
 
+    internal int Count => _frames?.Length ?? 0;
```
Comment (MackinnonBuck) [inline src/Components/Components/src/ParameterView.cs:49]: I don't think this is used anywhere, is it?

File: src/Components/Components/src/ResourceAssetProperty.cs [hunk 4]
```diff
@@ -0,0 +1,22 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+namespace Microsoft.AspNetCore.Components;
+
+/// <summary>
+/// A resource proeperty.
```
Comment (MackinnonBuck) [inline src/Components/Components/src/ResourceAssetProperty.cs:7]:
```suggestion
/// A resource property.
```

File: src/Components/Endpoints/src/Assets/ImportMap.cs [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    /// <summary>
+    /// Gets or sets the <see cref="HttpContext"/> for the component.
+    /// </summary>
+    [CascadingParameter] public HttpContext? HttpContext { get; set; } = null;
+
+    /// <summary>
+    /// Gets or sets the import map definition to use for the component. If not set
+    /// the component will generte the import map based on the assets defined for this
+    /// application.
+    /// </summary>
```
Comment (MackinnonBuck) [inline src/Components/Endpoints/src/Assets/ImportMap.cs:28]:
```suggestion
    /// <summary>
    /// Gets or sets the import map definition to use for the component. If not set,
    /// the component will generate the import map based on the assets defined for this
    /// application.
    /// </summary>
```

File: src/Components/Components/src/ResourceAssetCollection.cs [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+/// Describes a mapping of static assets to their corresponding unique URLs.
+/// </summary>
+public sealed class ResourceAssetCollection : IReadOnlyList<ResourceAsset>
+{
+    /// <summary>
+    /// An empty <see cref="ResourceAssetCollection"/>.
+    /// </summary>
+    public static readonly ResourceAssetCollection Empty = new([]);
+
+    private readonly Dictionary<string, ResourceAsset> _uniqueUrlMappings;
```
Comment (MackinnonBuck) [inline src/Components/Components/src/ResourceAssetCollection.cs:18]: I'd be curious if it's worth considering using a `FrozenDictionary` here if we want lookups to be as fast as possible.

File: src/Components/Endpoints/src/Builder/RazorComponentEndpointFactory.cs [hunk 7]
```diff
@@ -13,9 +13,9 @@
 
 namespace Microsoft.AspNetCore.Components.Endpoints;
 
-internal class RazorComponentEndpointFactory
+internal class RazorComponentEndpointFactory()
```
Comment (MackinnonBuck) [inline src/Components/Endpoints/src/Builder/RazorComponentEndpointFactory.cs:16]:
```suggestion
internal class RazorComponentEndpointFactory
```

File: src/Components/Endpoints/src/Builder/ResourceCollectionConvention.cs [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+using Microsoft.AspNetCore.Components.Web;
+
+namespace Microsoft.AspNetCore.Builder;
+
+internal class ResourceCollectionConvention(ResourceCollectionResolver resolver)
+{
+    string? _collectionUrl;
+    ImportMapDefinition? _collectionEndpointImportMap;
+    ResourceAssetCollection? _collection;
+    ImportMapDefinition? _collectionImportMap;
```
Comment (MackinnonBuck) [inline src/Components/Endpoints/src/Builder/ResourceCollectionConvention.cs:15]: Nit:
```suggestion
    private string? _collectionUrl;
    private ImportMapDefinition? _collectionEndpointImportMap;
    private ResourceAssetCollection? _collection;
    private ImportMapDefinition? _collectionImportMap;
```

File: src/Components/Endpoints/src/Builder/ResourceCollectionUrlEndpoint.cs [hunk 9]
```diff
@@ -0,0 +1,291 @@
...
+
+            builderConeg.Metadata.Add(encodingMetadata);
+
+            yield return fingerprintedGzBuilder;
+            yield return fingerprintedBuilder;
+            yield return fingerprintedBuilderConeg;
+            yield return gzBuilder;
+            yield return builderConeg;
+            yield return builder;
+        }
```
Comment (MackinnonBuck) [inline src/Components/Endpoints/src/Builder/ResourceCollectionUrlEndpoint.cs:289]: Why this and not:
```suggestion
            return [
                fingerprintedGzBuilder,
                fingerprintedBuilder,
                fingerprintedBuilderConeg,
                gzBuilder,
                builderConeg,
                builder,
            ];
        }
```
?

The former does generate quite a bit more code.

File: src/Components/test/E2ETest/Infrastructure/ServerFixtures/WebHostServerFixture.cs [hunk 10]
```diff
@@ -36,9 +36,15 @@ public override void Dispose()
...
+        try
+        {
+            // This can be null if creating the webhost throws, we don't want to throw here and hide
+            // the original exception.
+            await Host?.StopAsync();
+            Host?.Dispose();
+        }
+        catch
+        {
+        }
```
Comment (MackinnonBuck) [inline src/Components/test/E2ETest/Infrastructure/ServerFixtures/WebHostServerFixture.cs:48]: Temporary change?

File: src/Mvc/Mvc.Core/src/Builder/ControllerEndpointRouteBuilderExtensions.cs [hunk 11]
```diff
@@ -28,7 +28,13 @@ public static ControllerActionEndpointConventionBuilder MapControllers(this IEnd
 
         EnsureControllerServices(endpoints);
 
-        return GetOrCreateDataSource(endpoints).DefaultBuilder;
+        var result = GetOrCreateDataSource(endpoints).DefaultBuilder;
+        if (!result.Items.ContainsKey("__EndpointRouteBuilder"))
+        {
+            result.Items["__EndpointRouteBuilder"] = endpoints;
```
Comment (MackinnonBuck) [inline src/Mvc/Mvc.Core/src/Builder/ControllerEndpointRouteBuilderExtensions.cs:34]: Nit: This string literal is repeated multiple times

File: src/Mvc/Mvc.TagHelpers/src/LinkTagHelper.cs [hunk 12]
```diff
@@ -520,18 +523,71 @@ private void BuildLinkTag(string href, TagHelperAttributeList attributes, TagHel
...
-
+        hrefValue = GetVersionedResourceUrl(hrefValue);
         builder
             .AppendHtml(hrefName)
             .AppendHtml("=\"")
             .Append(hrefValue)
             .AppendHtml("\" ");
     }
 
+    private string GetVersionedResourceUrl(string url)
```
Comment (MackinnonBuck) [inline src/Mvc/Mvc.TagHelpers/src/LinkTagHelper.cs:534]: Is there a way this logic can be de-duped with the method of the same name in `ImageTagHelper`?

Comment (MackinnonBuck) [reply src/Mvc/Mvc.TagHelpers/src/LinkTagHelper.cs:534]: Same for `ScriptTagHelper`

File: src/Mvc/Mvc.ViewFeatures/src/Builder/ControllerActionEndpointConventionBuilderResourceCollectionExtensions.cs [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+    /// <returns></returns>
+    public static ControllerActionEndpointConventionBuilder WithStaticAssets(
+        this ControllerActionEndpointConventionBuilder builder,
+        string manifestPath = null)
+    {
+        ArgumentNullException.ThrowIfNull(builder);
+
+        if (builder.Items.TryGetValue("__EndpointRouteBuilder", out var endpointBuilder))
+        {
+            var (resolver, registered) = builder.Items.TryGetValue("__ResourceCollectionResolver", out var value)
```
Comment (MackinnonBuck) [inline src/Mvc/Mvc.ViewFeatures/src/Builder/ControllerActionEndpointConventionBuilderResourceCollectionExtensions.cs:31]: Nit: More repeated strings (not in the same file, but across projects).

Maybe this is something already on your TODO list, in which case I apologize if it's not helpful to comment on these!

File: src/StaticAssets/src/StaticAssetSelector.cs [hunk 14]
```diff
@@ -5,12 +5,29 @@
...
+
+    /// <summary>
+    /// The value associated to the selector and used to match against incoming requests.
+    /// </summary>
     public string Value { get; } = value;
+
+    /// <summary>
+    /// The static server quality associated to this selector. Used to break ties when a request matches multiple values
+    /// with the same degree of specificity.
+    /// </summary>
```
Comment (MackinnonBuck) [inline src/StaticAssets/src/StaticAssetSelector.cs:30]: Is this meant to say "static asset" quality?

**Review by MackinnonBuck — APPROVED** [review]

File: src/Mvc/Mvc.TagHelpers/src/LinkTagHelper.cs (hunk 12 above)
Comment (javiercn) [reply src/Mvc/Mvc.TagHelpers/src/LinkTagHelper.cs:532]: Done, shared the logic in an internal static helper method.

**Comment by Webreaper** [comment]
Out of interest (and since this ticket is somewhat related to hot reload) is any work being done in .Net 9 to make hot reload work reliably for anything other than the most simple projects? If there's an open issue, I'd really like to track it. We have significant problems with the reliability of hot reload in both VS and Rider, on any platform, once a hosted blazor server project gets past anything larger than a couple of trivial pages/components.

**Comment by javiercn** [comment]
@Webreaper this work is orthogonal to hot reload.

If you have concrete issues, it's best if you file issues for them and let us triage them. If there are already issues, then we'll dupe against those. With that said, the best way for us to resolve hot reload issues is if you include repro projects and detailed steps on how to trigger the concrete problems that you find. Without that, it's really hard for us to act on particular issues.

**Comment by Webreaper** [comment]
Thanks. The problem is that with hot reload, because it requires a substantially large and complex project to break it, a small repro is challenging. I can't post my enterprise blazor platform....

But I'll see what I can do and raise a separate issue. Thanks!
